File: 454.4Sum-II.py
# https://leetcode.com/problems/4sum-ii/
#
# algorithms
# Medium (49.45%)
# Total Accepted:    54,219
# Total Submissions: 109,642
# beats 89.26% of python submissions

from collections import defaultdict

class Solution(object):
    def fourSumCount(self, A, B, C, D):
        """
        :type A: List[int]
        :type B: List[int]
        :type C: List[int]
        :type D: List[int]
        :rtype: int
        """
        hash_map_a_b = defaultdict(int)

        for idx_A, m in enumerate(A):
            for idx_B, n in enumerate(B):
                s = m + n
                hash_map_a_b[s] += 1

        res = 0

        for m in C:
            for n in D:
                s = m + n
                if -s in hash_map_a_b:
                    res += hash_map_a_b[-s]

        return res

File: 454.4Sum-II.java
// https://leetcode.com/problems/4sum-ii/
//
// algorithms
// Medium (50.42%)
// Total Accepted:    63,990
// Total Submissions: 126,920


class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        int len = A.length;
        if (len == 0) {
            return 0;
        }
        
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < len; i++) {
            for (int j = 0; j < len; j++) {
                int s = A[i] + B[j];
                map.put(s, map.getOrDefault(s, 0) + 1);
            }
        }
        
        int res = 0;
        for (int i = 0; i < len; i++) {
            for (int j = 0; j < len; j++) {
                int s = C[i] + D[j];
                if (map.containsKey(-s)) {
                    res += map.get(-s);
                }
            }
        }
        
        return res;
    }
}